### PR TITLE
tests: always try adding a hotplug volume 3 times

### PR DESCRIPTION
**Pull request by acardace**
Try at least 3 times, this is done because `AddVolume` is inherently racy in the way it's implemented as when it patches the VM Status it expects the field `volumeRequests` to be there (by using a test json patch op), but at the same time virt-controller trims this field when all requests have been satified. To avoid hitting these we should explicitly try multiple times.

This is being changed after a few failures in the upstream CI where this race condition was seen
https://prow.ci.kubevirt.io/view/gcs/kubevirt-prow/logs/periodic-kubevirt-e2e-k8s-1.30-sig-storage/1775991881285505024.

Why isn't the `Eventually` not enough? Because it tries for 3 seconds and does not guarantee that the op will be tried more than once.

### What this PR does
Before this PR:
 Test [sig-storage] Hotplug [storage-req] Online VM should allow to hotplug 75 volumes simultaneously is flaky.

After this PR:
 Test [sig-storage] Hotplug [storage-req] Online VM should allow to hotplug 75 volumes simultaneously should not flake anymore.

Fixes #

### Why we need it and why it was done in this way
The following tradeoffs were made:

The following alternatives were considered:

Links to places where the discussion took place: 

### Checklist

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [ ] Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required
- [ ] PR: The PR description is expressive enough and will help future contributors
- [ ] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [ ] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [ ] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [ ] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least on e2e test
- [ ] Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.
- [ ] Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered

### Release note

```release-note
NONE
```

### CODE REVIEW

**Review by akalenyu — COMMENTED** [review]
> Because it tries for 3 seconds and does not guarantee that the op will be tried more than once.

Not sure I got that part. The timeout is 3 seconds with an interval of 1 second - why don't we try 3 times today?

**Comment by akalenyu** [comment]
> > > Because it tries for 3 seconds and does not guarantee that the op will be tried more than once.
> > 
> > 
> > Not sure I got that part. The timeout is 3 seconds with an interval of 1 second - why don't we try 3 times today?
> 
> If the whole op takes more than 3 seconds it just get tried once as the Eventually clause times out.
> 
> This is what I've seen happening by looking at virt-api and kube-api-server logs.

Gotcha

**Comment by akalenyu** [comment]
Looks good. Noticed some other uses of AddVolume, I guess those don't trigger the race as much as this particular one?
/lgtm
/cc @mhenriks

**Comment by acardace** [comment]
> AddVolume

This might be particularly prone to be racy since the nature of the test, I wouldn't touch the other usages just yet if you agree. First I want to see this test not being flaky anymore.

File: tests/storage/hotplug.go [hunk 1]
```diff
@@ -156,9 +156,19 @@ var _ = SIGDescribe("Hotplug", func() {
...
 	addVolumeVMWithSource := func(name, namespace string, volumeOptions *v1.AddVolumeOptions) {
-		Eventually(func() error {
-			return virtClient.VirtualMachine(namespace).AddVolume(context.Background(), name, volumeOptions)
-		}, 3*time.Second, 1*time.Second).ShouldNot(HaveOccurred())
+		var err error
+
+		// Try at least 3 times, this is done because `AddVolume` is inherently racy in the way it's implemented
+		// as when it patches the VM Status it expects the field `volumeRequests` to be there (by using a test json patch op),
+		// but at the same time virt-controller trims this field when all requests have been satified.
+		// To avoid hitting these we should explicitly try multiple times.
```
Comment (xpivarc) [inline tests/storage/hotplug.go:164]: +1

File: tests/storage/hotplug.go [hunk 2]
```diff
@@ -156,9 +156,19 @@ var _ = SIGDescribe("Hotplug", func() {
...
+		for i := 0; i < 3; i++ {
+			err = virtClient.VirtualMachine(namespace).AddVolume(context.Background(), name, volumeOptions)
+			if err == nil {
+				break
```
Comment (xpivarc) [inline tests/storage/hotplug.go:168]: nit: `break` -> `return`

**Review by xpivarc — COMMENTED** [review]
/approve

File: tests/storage/hotplug.go (hunk 2 above)
Comment (acardace) [reply tests/storage/hotplug.go:168]: if you do `return` you'd skip the `Expect`.

**Pull request merged**